src/networking.c

```diff
@@ -1590,8 +1590,99 @@ int _writeToClient(client *c, ssize_t *nwritten) {
         return C_OK;
     }
 
-    if (c->bufpos > 0) {
-        *nwritten = connWrite(c->conn,c->buf+c->sentlen,c->bufpos-c->sentlen);
+    int nbuf = listLength(c->reply);
+    if (nbuf > 0) {
+        if (c->bufpos > 0) ++nbuf; 
+        if (nbuf > 1) {
+            struct iovec iov[nbuf > IOV_MAX ? IOV_MAX : nbuf];
+            int iovcnt = 0;
+            size_t cum_bytes = 0;
+            /* If the successive buffer of responses is not empty, 
+             * add it to the buffer queue of writev() as well. */
+            if (c->bufpos > 0) {
+                iov[iovcnt].iov_base = c->buf + c->sentlen;
+                iov[iovcnt].iov_len = c->bufpos - c->sentlen;
+                cum_bytes += iov[iovcnt++].iov_len;
+            }
+            /* The first node of reply list might be incomplete from the last call,
+             * thus it needs to be calibrated to get the actual data address and length. */
+            size_t offset = c->bufpos > 0 ? 0 : c->sentlen;
+            listIter iter;
+            listNode *next;
+            clientReplyBlock *o;
+            listRewind(c->reply, &iter);
+            while ((next = listNext(&iter)) && iovcnt < IOV_MAX && cum_bytes < NET_MAX_WRITES_PER_EVENT) {
+                o = listNodeValue(next);
+                if (o->used == 0) { /* empty node, just release it and skip. */
+                    c->reply_bytes -= o->size;
+                    listDelNode(c->reply, next);
+                    continue;
+                }
+
+                iov[iovcnt].iov_base = o->buf + offset;
+                iov[iovcnt].iov_len = o->used - offset;
+                cum_bytes += iov[iovcnt++].iov_len;
+                offset = 0;
+            }
+            if (iovcnt == 0) return C_OK;
+            *nwritten = connWritev(c->conn, iov, iovcnt);
+            if (*nwritten <= 0) return C_ERR;
+
+            /* Locate the new node which has leftover data and
+            * release all nodes in front of it. */
+            ssize_t remaining = *nwritten;
+            if (c->bufpos > 0) { /* deal with successive buffer of responses first. */
```

Comment:
i don't know where you saw that buffer being refereed as "successive", IIRC it is called "static reply buffer"
```suggestion
            if (c->bufpos > 0) { /* deal with static response buffer first. */
```